support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/web/endpoints/OAuth20RevocationEndpointController.java

```diff
@@ -0,0 +1,191 @@
+            .map(String::valueOf).orElse(StringUtils.EMPTY);
+
+        if (isAsupportedTokenType(token)) {
+            return generateRevocationResponse(token, clientId, response);
+        }
+        return OAuth20Utils.writeError(response, OAuth20Constants.INVALID_REQUEST);
+    }
+
+    /**
+     * Generate revocation token response.
+     *
+     * @param token the token to revoke
+     * @param client the client who requests the revocation
+     * @param response the response
+     * @return the model and view
+     */
+    private ModelAndView generateRevocationResponse(final String token,
+                                                    final String client,
+                                                    final HttpServletResponse response) {
+
+        if (token.startsWith(OAuth20RefreshToken.PREFIX)) {
+            val registryToken = getOAuthConfigurationContext().getTicketRegistry().getTicket(token, OAuth20RefreshToken.class);
+
+            if (registryToken == null) {
+                LOGGER.error("Provided token [{}] has not been found in the ticket registry", token);
+            } else {
+                if (!StringUtils.equals(client, registryToken.getClientId())) {
+                    LOGGER.error("Provided token [{}] is not related with the service [{}]", token, client);
+                    return OAuth20Utils.writeError(response, OAuth20Constants.INVALID_REQUEST);
+                }
+                revokeToken(registryToken);
+            }
+        } else {
+            val registryToken = getOAuthConfigurationContext().getTicketRegistry().getTicket(token, OAuth20AccessToken.class);
+
+            if (registryToken == null) {
+                LOGGER.error("Provided token [{}] has not been found in the ticket registry", token);
+            } else {
+                if (!StringUtils.equals(client, registryToken.getClientId())) {
+                    LOGGER.error("Provided token [{}] is not related with the service [{}]", token, client);
+                    return OAuth20Utils.writeError(response, OAuth20Constants.INVALID_REQUEST);
+                }
+                revokeToken(registryToken);
+            }
+        }
+
+        val mv = new ModelAndView(new MappingJackson2JsonView());
+        mv.setStatus(HttpStatus.OK);
+        return mv;
+    }
+
+    /**
+     * Revoke the provided Refresh Token and it's related Access Tokens.
+     *
+     * @param token the token
+     * @return the model and view
+     */
+    private void revokeToken(final OAuth20RefreshToken token) {
+        LOGGER.debug("Revoking token [{}]", token);
+        getOAuthConfigurationContext().getTicketRegistry().deleteTicket(token);
+
+        token.getAccessTokens().forEach(item-> {
+            LOGGER.debug("Revoking Access Token [{}] related to Refresh Token [{}]", item, token);
+            getOAuthConfigurationContext().getTicketRegistry().deleteTicket(item);
```

Comment:
Remove log line, Call `revokeToken` instead in a lambda to avoid duplicate code.

Reply:
Done.